_posts/2025-04-22-test-classloading-rewrite.adoc

```diff
@@ -0,0 +1,60 @@
+
+Over time, test-related defects were building up that couldn't be changed without a fundamental rewrite of test classloading. The Quarkus test code itself was also growing ever-more complex as it tried to work around various JUnit edge cases. Moving test instances from one classloader to another involved serializing and deserialization, which is harder to implement on newer JVM versions with tighter class security. For example, Quarkus used to use XStream as the serialization provider, but XStream stopped working in Java 17.
+
+What if, instead, tests were simply run in the same classloader used to to load them?
+
+== What you need to do
+
+From Quarkus 3.22 onwards, this is exactly how `@QuarkusTest` classloading works.
+What do your tests need to change in order to work with the new architecture?
+Hopefully, nothing.
+One of the goals of this change was that the rewrite didn't touch any tests in our test suite, to make sure they'd all continue working without updates.
+In practice, there have been a few hiccups and we've discovered some edge cases in the broader ecosystem.
+
+=== Dropped support
+
+- *`@TestProfile` on `@Nested` tests.* Mixing different test profiles and test resources on `@Nested` tests is no longer supported. By definition, every `@TestProfile` must get its own Quarkus application and classloader. Having multiple classloaders execute one test isn't compatible with loading the test with the classloader used to run it.
+- *Version 2.x of the Maven Surefire plugin*. Versions below 3.x of the Maven Surefire plugin will no longer work with `@QuarkusTest`. Version 3 of the Surefire plugin was released in 2023, so version 2 is now rather old.
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause port conflicts and cross-talk on configuration values. We're hoping to have a https://github.com/quarkusio/quarkus/issues/45785[fix] for this in the next release. As a workaround, splitting conflicting tests into separate projects should fix symptoms.
+- *Config access from JUnit conditions*. Using a `ConfigProvider` from a custom JUnit condition will https://github.com/quarkusio/quarkus/issues/47081[trigger a `ServiceConfigurationError`]. The workaround is to set the thread context classloader to `this.getClass().getClassLoader()` before reading config, and then set it back afterwards.
+- Increased memory footprint running tests. For suites using multiple profiles and resources, more heap or metaspace may be needed.
```

Comment:
why? I would have expected less usage as there is just one classloader ...or is it because before there was one shared "heavy" classloader and many junit classloaders and now there are just many "heavy" junit classloaders ?

Reply:
The overall number of classloaders is the same, and they're basically the same size as before. The extra pressure on the heap happens because we have more of them present on the heap at the same time. It used to be we'd close one before creating the next. It's a side effect of frontloading the classloading.

Reply:
gotcha....but they will all get closed/collected per test class or suite? i assume they must be cleaned up as otherwise continues testing will have a hard time :)

Reply:
(We collect them as soon as we're done with the test/test profile group, same as before.)